doc/reference/controllers.md

```diff
@@ -96,46 +94,184 @@ rules:
-    mainPool:
-      servers:
-      - url: http://127.0.0.1:9095
-      loadBalance:
-        policy: roundRobin
+- name: validator 
+  kind: Validator
+  headers: 
+    X-ID:
+      values: ["user1", "user2"] 
+- name: proxy
+  kind: Proxy 
+  pools:
+  - servers:
+    - url: http://127.0.0.1:9095 
 ```
+In this case, if a request’s header doesn’t have the key `X-ID` or its value is not `user1` or `user2`, then the `validator` filter returns an` invalid` result and the pipeline jumps to `END`, the `proxy` filter does not execute. 
 
-| Name    | Type                                         | Description                          | Required |
-| ------- | -------------------------------------------- | ------------------------------------ | -------- |
-| flow    | [httppipeline.Flow](#httppipelineFlow)       | Flow of http pipeline                | No       |
-| Filters | [][httppipeline.Filter](#httppipelineFilter) | Filters definitions of http pipeline | Yes      |
+
+> `jumpIf` can only jump to filters behind the current filter.
+
+
+Resilience controls the behavior of backend filters, for now, it only works for filter `proxy`. There are two kinds of resilience, `Retry` and `CircuitBreak`. Check [resilience](#resilience) for more details. Following config adds retry resilience to pipeline: 
+```yaml
+name: http-pipeline-example3
+kind: Pipeline
+# END is built in filter name, it stops execution of pipeline and returns. 
+flow:
+- filter: proxy 
+
+filters:
+- name: proxy
+  kind: Proxy 
+  pools:
+  - servers:
+    - url: http://127.0.0.1:9095
+    retryPolicy: retry
+    
+resilience:
+- name: retry
+  kind: Retry
+  maxAttempts: 3
+``` 
+In this case, if `proxy` returns non-empty results, then resilience retry reruns the `proxy` filter until `proxy` returns empty results or gets the max attempts. 
+
+The `flow` also supports `namespace`, so the pipeline can support workflows that contain multiple requests and responses. 
+```yaml
+name: http-pipeline-example4
+kind: Pipeline
+flow: 
+- filter: validator
+  jumpIf:
+    invalid: END 
+- filter: requestBuilderFoo 
+  namespace: foo 
+- filter: proxyFoo 
+  namespace: foo 
+- filter: requestBuilderBar 
+  namespace: bar 
+- filter: proxyBar
+  namespace: bar 
+- filter: responseBuilder 
+
+filters: 
+- name: requestBuilder
+  kind: RequestBuilder
+  ...
+... 
+```
+In this case, `requestBuilderFoo` creates a request in namespace `foo`, and `proxyFoo` sends `foo` request and puts the response into namespace `foo`. `requestBuilderBar` creates a request in namespace `bar` and `proxyBar` sends `bar` request and puts the response into namespace `bar`. Finally, `requestBuilder` create a response by using responses in namespace `foo` and `bar` and putting it into the default namespace. 
+
+> If not set, the filter works in the default namespace `DEFAULT`.
+
+
+The `alias` in `flow` can give the filter an alias name.  So you can easily create different filters with the same spec. For example: 
```

Comment:
```suggestion
The `alias` in `flow` gives a filter an alias to help re-use the filter so that we can use the alias to distinguish each of its appearances in the flow.
```